pandas/tests/indexes/multi/test_constructor.py

```diff
@@ -472,3 +472,83 @@ def test_from_tuples_with_tuple_label():
     idx = pd.MultiIndex.from_tuples([(2, 1), (4, (1, 2))], names=('a', 'b'))
     result = pd.DataFrame([2, 3], columns=['c'], index=idx)
     tm.assert_frame_equal(expected, result)
+
+
+def test_from_frame():
+    df = pd.DataFrame([['a', 'a'], ['a', 'b'], ['b', 'a'], ['b', 'b']],
+                      columns=['L1', 'L2'])
+    expected = pd.MultiIndex.from_tuples([('a', 'a'), ('a', 'b'),
+                                          ('b', 'a'), ('b', 'b')],
+                                         names=['L1', 'L2'])
+    result = pd.MultiIndex.from_frame(df)
+    tm.assert_index_equal(expected, result)
+
+
+@pytest.mark.parametrize('squeeze,input_type,expected', [
+    (True, 'multi', pd.MultiIndex.from_tuples([('a', 'a'), ('a', 'b'),
+                                               ('b', 'a'), ('b', 'b')],
+                                              names=['L1', 'L2'])),
+    (True, 'single', pd.Index(['a', 'a', 'b', 'b'], name='L1')),
+    (False, 'multi', pd.MultiIndex.from_tuples([('a', 'a'), ('a', 'b'),
+                                                ('b', 'a'), ('b', 'b')],
+                                               names=['L1', 'L2'])),
+    (False, 'single', pd.MultiIndex.from_tuples([('a',), ('a',),
+                                                 ('b',), ('b',)],
+                                                names=['L1']))
+])
+def test_from_frame_squeeze(squeeze, input_type, expected):
+    if input_type == 'multi':
+        df = pd.DataFrame([['a', 'a'], ['a', 'b'], ['b', 'a'], ['b', 'b']],
+                          columns=['L1', 'L2'])
+    elif input_type == 'single':
+        df = pd.DataFrame([['a'], ['a'], ['b'], ['b']], columns=['L1'])
+
+    result = pd.MultiIndex.from_frame(df, squeeze=squeeze)
+    tm.assert_index_equal(expected, result)
+
+
+def test_from_frame_non_frame():
+    with tm.assert_raises_regex(TypeError, 'Input must be a DataFrame'):
+        pd.MultiIndex.from_frame([1, 2, 3, 4])
+
+
+def test_from_frame_dtype_fidelity():
+    df = pd.DataFrame({
+        'dates': pd.date_range('19910905', periods=6, tz='US/Eastern'),
+        'a': [1, 1, 1, 2, 2, 2],
+        'b': pd.Categorical(['a', 'a', 'b', 'b', 'c', 'c'], ordered=True),
+        'c': ['x', 'x', 'y', 'z', 'x', 'y']
+    })
+    original_dtypes = df.dtypes.to_dict()
+
+    expected_mi = pd.MultiIndex.from_arrays([
+        pd.date_range('19910905', periods=6, tz='US/Eastern'),
+        [1, 1, 1, 2, 2, 2],
+        pd.Categorical(['a', 'a', 'b', 'b', 'c', 'c'], ordered=True),
+        ['x', 'x', 'y', 'z', 'x', 'y']
+    ], names=['dates', 'a', 'b', 'c'])
+    mi = pd.MultiIndex.from_frame(df)
+    mi_dtypes = {name: mi.levels[i].dtype for i, name in enumerate(mi.names)}
+
+    tm.assert_index_equal(expected_mi, mi)
+    assert original_dtypes == mi_dtypes
+
+
+@pytest.mark.parametrize('names_in,names_out', [
```

Comment:
where is the test about guaranteed sort order?

Reply:
Added, this also raised a minor bug in to_frame. If you have a column name 0 in your columns not in the 0th position, the name will be replaced with its positional index. I have updated to_frame to account for this.

Reply:
consider these 2 examples:

    >>> pd.MultiIndex.from_arrays([[1,2,3],[1,2,3],[1,2,3]], names=[2,1,0]).to_frame()
    >>> pd.MultiIndex.from_arrays([[1,2,3],[1,2,3],[1,2,3]], names=['a',0,'b']).to_frame()

This is due to the `(name or level)` key in the dict comprehension.
I'd be surprised if anyone ever ran into this, but the update to to_frame handles this now.